src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs

```diff
@@ -306,9 +307,48 @@ private async Task ProcessNegotiate(HttpContext context, HttpConnectionDispatche
             }
         }
 
-        private static void WriteNegotiatePayload(IBufferWriter<byte> writer, string connectionId, HttpContext context, HttpConnectionDispatcherOptions options)
+        private void WriteNegotiatePayload(IBufferWriter<byte> writer, string connectionId, HttpContext context, HttpConnectionDispatcherOptions options)
         {
             var response = new NegotiationResponse();
+
+            if (context.Request.Query.TryGetValue("NegotiateVersion", out var queryStringVersion))
+            {
+                // Set the negotiate response to the protocol we use.
+                var queryStringVersionValue = queryStringVersion.ToString();
+                if (int.TryParse(queryStringVersionValue, out var clientProtocolVersion))
+                {
+                    if (clientProtocolVersion < options.MinimumProtocolVersion)
+                    {
+                        response.Error = $"The client requested version '{clientProtocolVersion}', but the server does not support this version.";
+                        Log.NegotiateProtocolVersionMismatch(_logger, clientProtocolVersion);
+                        NegotiateProtocol.WriteResponse(response, writer);
+                        return;
+                    }
+                    else if (clientProtocolVersion > _protocolVersion)
+                    {
+                        response.Version = _protocolVersion;
+                    }
+                    else
+                    {
+                        response.Version = clientProtocolVersion;
+                    }
+                }
+                else
+                {
+                    response.Error = $"The client requested an invalid protocol version '{queryStringVersionValue}'";
+                    Log.InvalidNegotiateProtocolVersion(_logger, queryStringVersionValue);
+                    NegotiateProtocol.WriteResponse(response, writer);
+                    return;
+                }
+            }
+            else if (options.MinimumProtocolVersion > 0)
+            {
+                // NegotiateVersion wasn't parsed meaning the client requests version 0.
+                response.Error = $"The client requested version '0', but the server does not support this version.";
```

Comment:
And a test for this

Reply:
Preferably one of these tests would verify there is no connection ID in the response payload